File: go-backend-interview-prep/01_go_basics/04_constants/main.go
package main

import "fmt"

// Simple constants
const appVersion = "1.0.0"
const maxConnections int = 100

// Enum-style with iota
const (
	StatusPending  = iota // 0
	StatusActive          // 1
	StatusInactive        // 2
	StatusDeleted         // 3
)

// iota with bit shifts (useful for file permissions, sizes, etc.)
const (
	KB = 1 << (10 * (iota + 1)) // 1 << 10 = 1024
	MB                           // 1 << 20 = 1048576
	GB                           // 1 << 30 = 1073741824
)

// iota resets in each new const block
const (
	Red   = iota // 0 (reset)
	Green        // 1
	Blue         // 2
)

func main() {
	// --- Simple constants ---
	fmt.Println("version:", appVersion)
	fmt.Println("max connections:", maxConnections)

	// --- Enum constants ---
	fmt.Println("--- Status Enum ---")
	fmt.Println("Pending:", StatusPending)
	fmt.Println("Active:", StatusActive)
	fmt.Println("Inactive:", StatusInactive)
	fmt.Println("Deleted:", StatusDeleted)

	// --- Bit-shift constants ---
	fmt.Println("--- Sizes ---")
	fmt.Println("KB:", KB)
	fmt.Println("MB:", MB)
	fmt.Println("GB:", GB)

	// --- iota reset proof ---
	fmt.Println("--- Colors (iota reset) ---")
	fmt.Println("Red:", Red)
	fmt.Println("Green:", Green)
	fmt.Println("Blue:", Blue)

	// --- Untyped constant flexibility ---
	// 'pi' has no fixed type, so it works with float32 and float64
	const pi = 3.14159
	var f32 float32 = pi
	var f64 float64 = pi
	fmt.Println("float32 pi:", f32)
	fmt.Println("float64 pi:", f64)
}

File: go-backend-interview-prep/01_go_basics/04_constants/main.py
# Python equivalent of Constants (compare with main.go)
# Python has no true const keyword. Convention: UPPER_CASE names.

from enum import IntEnum

# "Constants" by convention (nothing prevents reassignment)
APP_VERSION = "1.0.0"
MAX_CONNECTIONS = 100


# Go iota equivalent: use IntEnum
class Status(IntEnum):
    PENDING = 0
    ACTIVE = 1
    INACTIVE = 2
    DELETED = 3


# Bit-shift constants
KB = 1 << 10
MB = 1 << 20
GB = 1 << 30


# Another enum (iota resets in Go; separate class in Python)
class Color(IntEnum):
    RED = 0
    GREEN = 1
    BLUE = 2


def main():
    # --- Simple constants ---
    print("version:", APP_VERSION)
    print("max connections:", MAX_CONNECTIONS)

    # --- Enum constants ---
    print("--- Status Enum ---")
    print("Pending:", Status.PENDING)
    print("Active:", Status.ACTIVE)
    print("Inactive:", Status.INACTIVE)
    print("Deleted:", Status.DELETED)

    # --- Bit-shift constants ---
    print("--- Sizes ---")
    print("KB:", KB)
    print("MB:", MB)
    print("GB:", GB)

    # --- Color enum ---
    print("--- Colors ---")
    print("Red:", Color.RED)
    print("Green:", Color.GREEN)
    print("Blue:", Color.BLUE)

    # --- Python has no untyped constant flexibility ---
    # In Go, untyped const adapts to context. Python just has one float type.
    PI = 3.14159
    print("PI:", PI)


if __name__ == "__main__":
    main()
